File: hyper-collate-core/src/main/java/nl/knaw/huygens/hypercollate/model/TextNode.java
package nl.knaw.huygens.hypercollate.model;

/*-
 * #%L
 * hyper-collate-core
 * =======
 * Copyright (C) 2017 - 2018 Huygens ING (KNAW)
 * =======
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import eu.interedition.collatex.Token;
import static java.util.stream.Collectors.joining;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class TextNode implements Node {
  final Map<String, Token> tokenMap = new HashMap<>();
  private final Map<String, List<Integer>> branchPaths = new HashMap<>();
  public static final String LABEL = "Text";

  TextNode(Token... tokens) {
    for (Token token : tokens) {
      addToken(token);
    }
  }

  public void addToken(Token token) {
    if (token != null && token.getWitness() != null) {
      tokenMap.put(token.getWitness().getSigil(), token);
    }
  }

  public Token getTokenForWitness(String sigil) {
    return tokenMap.get(sigil);
  }

  public Set<String> getSigils() {
    return tokenMap.keySet();
  }

  @Override
  public String toString() {
    String tokensString = getSigils()//
        .stream()//
        .sorted()//
        .map(tokenMap::get)//
        .map(Token::toString)//
        .collect(joining(", "));
    return "(" + tokensString + ")";
  }

  public List<Integer> getBranchPath(String s) {
    return branchPaths.get(s);
  }

  public void addBranchPath(String sigil, List<Integer> branchPath) {
    branchPaths.put(sigil, branchPath);
  }
}

File: hyper-collate-core/src/main/kotlin/nl/knaw/huygens/hypercollate/model/TextNode.kt
package nl.knaw.huygens.hypercollate.model

/*-
* #%L
 * hyper-collate-core
 * =======
 * Copyright (C) 2017 - 2020 Huygens ING (KNAW)
 * =======
 * Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *  
 *       http://www.apache.org/licenses/LICENSE-2.0
 *  
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 * #L%
*/

import eu.interedition.collatex.Token
import java.util.*
import java.util.stream.Collectors

open class TextNode(vararg tokens: Token) : Node {
    private val tokenMap: MutableMap<String, Token> = HashMap()
    private val branchPaths: MutableMap<String, List<Int>> = HashMap()

    fun addToken(token: Token) {
        if (token.witness != null) {
            tokenMap[token.witness.sigil] = token
        }
    }

    fun tokenForSigil(sigil: String): Token? =
            tokenMap[sigil]

    val sigils: Set<String>
        get() = tokenMap.keys

    override fun toString(): String {
        val tokensString = sigils.stream()
                .sorted()
                .map { key: String -> tokenMap[key]!! }
                .map { obj: Token -> obj.toString() }
                .collect(Collectors.joining(", "))
        return "($tokensString)"
    }

    fun branchPathForSigil(s: String): List<Int> =
            branchPaths[s]!!

    fun addBranchPath(sigil: String, branchPath: List<Int>) {
        branchPaths[sigil] = branchPath
    }

    companion object {
        const val LABEL = "Text"
    }

    init {
        for (token in tokens) {
            addToken(token)
        }
    }
}
